onnx/onnx.proto

```diff
@@ -212,6 +237,44 @@ message TensorProto {
 message SparseTensorProto {
   // The dimensions in the tensor.
   repeated int64 dims = 1;
+  // This field MUST be present this version of the IR.
   optional TensorProto indices = 2;
+  // This field MUST be present this version of the IR.  
   optional TensorProto values = 3;
 }
+
+// Define the types.
+message TypeProto {
+  message TensorTypeProto {
+  // This field MUST NOT have the value of UNDEFINED
+  // This field MUST be present for this version of the IR.
+    optional TensorProto.DataType elem_type = 1;
+  }
+
+  message SparseTensorTypeProto {
+  // This field MUST NOT have the value of UNDEFINED
+  // This field MUST be present for this version of the IR.
+    optional TensorProto.DataType elem_type = 1;
+  }
+
+  oneof value {
+    // The type of a tensor.
+    TensorTypeProto tensor_type = 1;
+
+    // The type of a sparse tensor.
+    SparseTensorTypeProto sparse_tensor_type = 2;
+  }
+}
+
+// Defines a tensor shape. A dimension can be either an integer value
+// or a symbolic variable. A symbolic variable represents an unknown
+// dimension.
+message TensorShapeProto {
```

Comment:
another (less descriptive) option would be to use negative values in "shape" as symbolic references, e.g. "-1", "-2", etc. It'd be less descriptive but would simplify the format and use the same `repeated int64 dims` for both fully-realized shape and partially specified shape.

Reply:
That would work, however using symbolic names has two advantages:

1. It coins a name one can refer to in doc_string.

2. If the same name appears two or more times on a given graph or node, it's value must be the same, which addresses correlating dimensions across input and output parameters..

Reply:
Agree, symbolic names are indeed more readable. Probably it's fine at expense of protobuf brevity.

Reply:
It could also be used for more efficient memory reuse, when we for example statically plan for the memory allocation/reuse.

Reply:
Yes, using negative numbers would work, but somewhat less elegant in my opinion.